### PR TITLE
[Fixes #35] DEPRECATION constructor/of broken

### LINKED ISSUE #35: fp-constructor/of broken
fp-constructor is implemented wrong. Probably my fault.

1. `.constructor` should be an instance method (on the instance's delegate prototype), not a static property. It should refer to the specific stamp which created the instance. (See [type representatives](https://github.com/fantasyland/fantasy-land#type-representatives))
2. `.of()` should be a static prop (on the stamp). It should also refer to the stamp. (See [.of() method](https://github.com/fantasyland/fantasy-land#of-method))

In other words, it should create the stamp equivalent to the following factory:

```js
const createUser = ({
  userName = 'Anonymous',
  avatar = 'anon.png'
} = {}) => Object.assign(
  Object.create({
    constructor: createUser
  }), {
    userName,
    avatar
  }
);

createUser.of = createUser;


// testing .of and .constructor:
const empty = ({ constructor } = {}) => constructor.of ?
  constructor.of() :
  undefined
;

const foo = createUser({ userName: 'Empty', avatar: 'me.png' });

console.log(
  empty(foo), // { avatar: "anon.png", userName: "Anonymous" }
  foo.constructor === createUser.of, // true
  createUser.of === createUser      // true
);
```

`.constructor` and `.of()` can be used as a tool to create Functors and Monads, to derive operations for algebraic datatypes, etc...

Incidentally, `.constructor` and `.of()` go hand-in-hand. If you want one, you almost certainly want both.

### PR DESCRIPTION
**Pull request by ericelliott**
See #35

### CODE REVIEW

**Comment by koresar** [comment]
First question. Why the name of the PR states ** DEPRECATION**? What are you deprecating?

**Comment by koresar** [comment]
This repo does not do package lock (yet). We might need it in the future. As soon as it is needed I'll add `package-lock` files to every package.
Please, remove the file.

File: packages/fp-constructor/index.js [hunk 1]
```diff
@@ -2,6 +2,10 @@ var compose = require('@stamp/compose');
 
 module.exports = compose({
   composers: [function (opts) {
+    opts.stamp.of = opts.stamp;
     opts.stamp.constructor = opts.stamp;
+    opts.stamp.compose.methods = Object.assign({}, opts.stamp.compose.methods, {
```
Comment (koresar) [inline packages/fp-constructor/index.js]: There is no benefit in composers to create new (often large) object. Composers were designed to mutate stamps. Please, remove new object creation. This should suffice:
```js
opts.stamp.compose.methods.constructor = opts.stamp;
```

Comment (koresar) [inline packages/fp-constructor/index.js]: Oh, wait. The `.methods` could be missing.
```js
if (stamp.compose.methods) stamp.compose.methods.constructor = stamp;
else stamp.compose.methods = {constructor: stamp};
```

**Comment by ericelliott** [comment]
> First question. Why the name of the PR states ** DEPRECATION**? What are you deprecating?

`Stamp.constructor` was created in error. That was never the intended behavior. It's being deprecated in favor of `Stamp.of`. The `.constructor` property is an instance prototype property.

> This repo does not do package lock (yet). We might need it in the future. As soon as it is needed 
 I'll add package-lock files to every package.
> Please, remove the file.

Sure. We'll need to add it to `.gitignore` -- latest npm adds the lock file automatically.

**Comment by koresar** [comment]
Thanks! Looks good.
[TIL](https://docs.npmjs.com/files/package-lock.json):
> One key detail about package-lock.json is that it cannot be published

After publishing this package the [README](https://github.com/stampit-org/stamp/tree/master/packages/fp-constructor) will become untrue. Could you please update it the way you see it?

**Comment by ericelliott** [comment]
Check the docs. Lots more detail on why this package is useful. =)

**Comment by koresar** [comment]
I wish I had this kind of writing talent 👍

**Comment by koresar** [comment]
I'm going to publish as v0.2.

In few days I'm going to publish all the packages as v1.0. This one included. Will remove the deprecated function.

**Comment by koresar** [comment]
v0.2 published: https://www.npmjs.com/package/@stamp/fp-constructor

**Comment by koresar** [comment]
I'd be very happy to see how you use this package. A blog post, or a repo link would be fine. I'm very interested.

**Comment by ericelliott** [comment]
Start with [Functors and categories](https://medium.com/javascript-scene/functors-categories-61e031bac53f)

**Pull request merged**